winpr/libwinpr/sspi/Kerberos/kerberos.c

```diff
@@ -659,117 +659,196 @@ static BOOL kerberos_mk_tgt_token(SecBuffer* buf, int msg_type, char* sname, cha
 	return ret;
 }
 
-static BOOL kerberos_rd_tgt_token(const sspi_gss_data* token, char** target, krb5_data* ticket)
+static size_t append(char* dst, size_t dstSize, const char* src)
 {
-	WinPrAsn1Decoder dec;
-	WinPrAsn1Decoder dec2;
-	BOOL error = 0;
-	WinPrAsn1_tagId tag = 0;
+	const size_t dlen = strnlen(dst, dstSize);
+	const size_t slen = strlen(src);
+	if (dlen + slen >= dstSize)
+		return SIZE_MAX;
+	if (!strncat(dst, src, slen))
+		return SIZE_MAX;
+	return dlen + slen;
+}
+
+static BOOL kerberos_rd_tgt_req_tag2(WinPrAsn1Decoder* dec, char* buf, size_t len, size_t* poffset)
+{
+	BOOL rc = FALSE;
+	WinPrAsn1Decoder seq = { 0 };
+	size_t offset = *poffset;
+
+	/* server-name [2] PrincipalName (SEQUENCE) */
+	if (!WinPrAsn1DecReadSequence(dec, &seq))
+		goto end;
+
+	/* name-type [0] INTEGER */
+	BOOL error = FALSE;
 	WinPrAsn1_INTEGER val = 0;
-	size_t len = 0;
-	wStream s;
-	char* buf = NULL;
-	char* str = NULL;
+	if (!WinPrAsn1DecReadContextualInteger(&seq, 0, &error, &val))
+		goto end;
 
-	WINPR_ASSERT(token);
+	/* name-string [1] SEQUENCE OF GeneralString */
+	if (!WinPrAsn1DecReadContextualSequence(&seq, 1, &error, dec))
+		goto end;
 
-	WinPrAsn1Decoder_InitMem(&dec, WINPR_ASN1_DER, (BYTE*)token->data, token->length);
+	WinPrAsn1_tag tag = 0;
+	while (WinPrAsn1DecPeekTag(dec, &tag))
+	{
+		char* lstr = NULL;
+		if (!WinPrAsn1DecReadGeneralString(dec, &lstr))
+			goto end;
 
-	/* KERB-TGT-REQUEST (SEQUENCE) */
-	if (!WinPrAsn1DecReadSequence(&dec, &dec2))
+		if (buf != &buf[offset])
+		{
+			offset = append(buf, len, "/");
+			if (offset > len)
+				goto end;
+		}
+
+		offset = append(buf, len, lstr);
+		if (offset > len)
+			goto end;
+		free(lstr);
+	}
+
+	rc = TRUE;
+end:
+	*poffset = offset;
+	return rc;
+}
+
+static BOOL kerberos_rd_tgt_req_tag3(WinPrAsn1Decoder* dec, char* buf, size_t len, size_t* poffset)
```

Comment:
The `offset` parameter here and in `kerberos_rd_tgt_req_tag2` seems to be redundant. It isn't really used in the functions themselves and the returned offset is discarded as well.